### ISSUE TITLE
Incorrect error message when server loader throws an error and clientLoader is present

### ISSUE THREAD

**Issue by justinwaite** [remix-run/remix#8300]
### Reproduction

https://stackblitz.com/edit/remix-run-remix-uyj9e9?file=app%2Froutes%2F_index.tsx

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 9.4.2 - /usr/local/bin/npm
    pnpm: 8.10.5 - /usr/local/bin/pnpm
  npmPackages:
    @remix-run/css-bundle: 2.4.0 => 2.4.0 
    @remix-run/dev: 2.4.0 => 2.4.0 
    @remix-run/node: 2.4.0 => 2.4.0 
    @remix-run/react: 2.4.0 => 2.4.0 
    @remix-run/serve: 2.4.0 => 2.4.0
```

### Used Package Manager

npm

### Expected Behavior

When an error is thrown in a server loader, I expect to be able to capture that error message and display it to the user in an error boundary. The error message in the reproduction that should be displayed is:
```
This is an error message that should render in the boundary.
```

### Actual Behavior

I get an initial flash of the expected error message, but quickly overridden by:
```
You defined a loader for route "routes/_index" but didn't return anything from your `loader` function. Please return a value or `null`.
```

If you remove the clientLoader, this behavior goes away.

Note: if you are looking at the stackblitz and see the "correct" error message, just refresh the app and most of the time you will see the "incorrect" error message as described above.

**Comment by justinwaite** [comment]
After some further investigation, it seems the problem is that calling `serverLoader` in the `clientLoader` function returns undefined in the event of `loader` throwing an error. Then `undefined` gets returned from `clientLoader`. I would expect that if `loader` throws an error, then calling `serverLoader` would likewise throw an error instead of undefined.

**Comment by brophdawg11** [comment]
Thanks for the reproduction and investigation!

**Comment by brophdawg11** [comment]
Resolved by #8304 and will be in the next release 👍

**Comment by justinwaite** [comment]
@brophdawg11 you guys are lightning fast. Thanks!
